### ISSUE TITLE
regrtest can reports false SUCCESS and skip tests when a -j worker dies

### ISSUE THREAD

**Issue by gpshead** [python/cpython#101634]
See in https://buildbot.python.org/all/#/builders/424/builds/3407/steps/5/logs/stdio (Raspbian)

```
...
0:04:46 load avg: 7.28 [ 48/434] test_isinstance passed -- running: test_largefile (2 min 54 sec)
0:05:03 load avg: 7.39 [ 49/434] test_utf8source passed -- running: test_largefile (3 min 11 sec), test_unittest (38.7 sec)
Warning -- regrtest worker thread failed: Traceback (most recent call last):
Warning --   File "/var/lib/buildbot/workers/3.x.gps-raspbian.nondebug/build/Lib/test/libregrtest/runtest_mp.py", line 342, in run
Warning --     mp_result = self._runtest(test_name)
Warning --                 ^^^^^^^^^^^^^^^^^^^^^^^^
Warning --   File "/var/lib/buildbot/workers/3.x.gps-raspbian.nondebug/build/Lib/test/libregrtest/runtest_mp.py", line 301, in _runtest
Warning --     stdout = stdout_fh.read().strip()
Warning --              ^^^^^^^^^^^^^^^^
Warning --   File "<frozen codecs>", line 322, in decode
Warning -- UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa9 in position 35: invalid start byte
Kill <TestWorkerProcess #1 running test=test_tempfile pid=7836 time=21.8 sec> process group
Kill <TestWorkerProcess #2 running test=test_largefile pid=7700 time=3 min 32 sec> process group

== Tests result: SUCCESS ==

385 tests omitted:
    test___all__ test__locale test__opcode test__osx_support
    test__xxinterpchannels test__xxsubinterpreters test_abc test_aifc
    test_argparse test_array test_asdl_parser test_ast test_asyncgen
    test_asyncio test_atexit test_audioop test_audit test_base64
...
```

This might be an infrastructure issue or an actual issue with real non utf-8 data making it through the mp sockets. Regardless, it should've triggered an error and exit 1 instead of a false SUCCESS exit 0.

### Linked PRs
* gh-106169
* gh-106174
* gh-106175

**Comment by vstinner** [comment]
I wrote PR #106169 to fix this bug.

**Comment by hugovk** [comment]
Thanks!

**Comment by vstinner** [comment]
The fix will not be backported to Python 3.11 and 3.12 branches: see https://github.com/python/cpython/pull/111950#issuecomment-1808004559
